console-subscriber/tests/support/subscriber.rs

```diff
@@ -0,0 +1,323 @@
+        .await
+    {
+        Ok(stream) => stream.into_inner(),
+        Err(err) => panic!("Client cannot connect to watch updates: {err}"),
+    };
+
+    let mut tasks = HashMap::new();
+
+    let signal_task = ExpectedTask::default().match_name(END_SIGNAL_TASK_NAME.into());
+    let mut signal_task_read = false;
+    while let Some(update) = stream.next().await {
+        let update = update.expect("update stream error");
+
+        if let Some(task_update) = &update.task_update {
+            for new_task in &task_update.new_tasks {
+                let mut actual_task = match new_task.id {
+                    Some(id) => ActualTask::new(id.id),
+                    None => continue,
+                };
+                let name = new_task
+                    .fields
+                    .iter()
+                    .find_map(|field| match field.name.as_ref()? {
+                        console_api::field::Name::StrName(name) if name == "task.name" => {
+                            Some(field.value.clone())
+                        }
+                        _ => None,
+                    })
+                    .flatten();
+                actual_task.name = match name {
+                    Some(Value::DebugVal(value)) => Some(value.clone()),
+                    Some(Value::StrVal(value)) => Some(value.clone()),
+                    _ => None, // Anything that isn't string-like shouldn't be used as a name.
+                };
+
+                if signal_task.matches_actual_task(&actual_task) {
+                    signal_task_read = true;
+                } else {
+                    tasks.insert(actual_task.id, actual_task);
+                }
+            }
+
+            for (id, stats) in &task_update.stats_update {
+                if let Some(task) = tasks.get_mut(id) {
+                    task.wakes = stats.wakes;
+                    task.self_wakes = stats.self_wakes;
+                }
+            }
+        }
+
+        if test_state.is_step(TestStep::TestFinished) && signal_task_read {
+            // Once the test finishes running and we've read the signal task, the test ends.
+            break;
+        }
+    }
+
+    tasks.into_values().collect()
+}
+
+/// Validate the expected tasks against the actual tasks.
+///
+/// Each expected task is checked in turn.
+///
+/// A matching actual task is searched for. If one is found it, the
+/// expected task is validated against the actual task.
+///
+/// Any validation errors result in failure. If no matches
+fn validate_expected_tasks(
+    expected_tasks: Vec<ExpectedTask>,
+    actual_tasks: Vec<ActualTask>,
+) -> Result<(), TestFailure> {
+    let failures: Vec<_> = expected_tasks
+        .iter()
+        .map(|expected| validate_expected_task(expected, &actual_tasks))
+        .filter_map(|r| match r {
+            Ok(_) => None,
+            Err(validation_error) => Some(validation_error),
+        })
```

Comment:
very tiny nit: there's actually a [method on `Result`](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.err) that just does this:
```suggestion
        .filter_map(Result::err)
```

Reply:
Nice! Hadn't used that before.